Installation failing on 1.6.0 and 1.6.1 in CI/GitHub Actions due to "error: missing initializer for field" compiler error

**Current behaviour**

Reported internally by the https://github.com/DataDog/datadog-api-client-ruby team. Their CI started failing with compilation errors:

* `collectors_stack.c:151:7: error: missing initializer for field ‘start_line’ of   ‘ddog_Function’ {aka ‘struct ddog_Function’}`
* `collectors_stack.c:307:5: error: missing initializer for field ‘start_line’ of
  ‘ddog_Function’ {aka ‘struct ddog_Function’}`
* `collectors_stack.c:344:5: error: missing initializer for field ‘start_line’ of
  ‘ddog_Function’ {aka ‘struct ddog_Function’}`
* `collectors_stack.c:376:5: error: missing initializer for field ‘is_folded’ of
  ‘ddog_Location’ {aka ‘struct ddog_Location’}`

This is caused by a combination of https://github.com/DataDog/dd-trace-rb/pull/2358 triggering erroneously (e.g. outside of ddtrace CI) and compiler warnings which we hadn't previously seen.

Specifically, installing ddtrace with the `CI=true` environment variable set will fail if the C compiler emits a warning. This is the case with `gcc version 9.4.0` (Ubuntu 20.04 default).

Issue can be worked around by defining `DD_PROFILING_NO_EXTENSION` environment variable to `true` prior to running `bundle install`. See https://github.com/DataDog/datadog-api-client-ruby/pull/1137 for an example.

**Expected behaviour**

A compiler warning should not cause a user of ddtrace to see an installation error. No workaround should be needed.

**Steps to reproduce**

(Still working on this)

**Environment**

* **ddtrace version:** 1.6.0 and 1.6.1
* **Configuration block (`Datadog.configure ...`):** (Not relevant)
* **Ruby version:** ruby 2.7.6p219 (2022-04-12 revision c9c2245c0a) [x86_64-linux]
* **Operating system:** Linux (GitHub Actions)
* **Relevant library versions:** (None)

Update: I was able to reproduce this locally using a `ubuntu:20.04` docker image and its default gcc (`gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.1)`).

This warning seems to be a GCC bug.

Test case:
```c
#include <stdio.h>
#include <stdint.h>

typedef struct ddog_Slice_c_char {
  const char *ptr;
  uintptr_t len;
} ddog_Slice_c_char;

typedef struct ddog_Slice_c_char ddog_CharSlice;

typedef struct ddog_Function {
  ddog_CharSlice name;
  int64_t start_line;
} ddog_Function;

typedef struct ddog_Line {
  struct ddog_Function function;
} ddog_Line;

int main(__attribute__((unused)) int argc, __attribute__((unused)) char **argv) {
  ddog_Function functions[1];

  functions[0] = (ddog_Function) {
      .name = (ddog_CharSlice) { .ptr = "potato", .len = sizeof("potato") - 1 }
    };

  return functions[0].start_line;
}
```

Compilation failure:

```
# gcc testcase.c -Wall -Wextra -Werror
testcase.c: In function 'main':
testcase.c:25:5: error: missing initializer for field 'start_line' of 'ddog_Function' {aka 'struct ddog_Function'} [-Werror=missing-field-initializers]
   25 |     };
      |     ^
testcase.c:13:11: note: 'start_line' declared here
   13 |   int64_t start_line;
      |           ^~~~~~~~~~
cc1: all warnings being treated as errors
```

But if we change the testcase as follows

```diff
   ddog_Function functions[1];

   functions[0] = (ddog_Function) {
-      .name = (ddog_CharSlice) { .ptr = "potato", .len = sizeof("potato") - 1 }
+      .name = { .ptr = "potato", .len = sizeof("potato") - 1 }
     };

   return functions[0].start_line;
```

The warning is gone.

Furthermore, this is not the first time we were bit by bugs around this area https://github.com/DataDog/dd-trace-rb/commit/b26f07df369cc94287d2d39b64b7811296d14c31, although modern versions of gcc seem fine.

---

The warning itself does not look like something to be fixed, since it doesn't even happen on more modern compilers. We could add a workaround to avoid it, but in this case I don't think the readability trade-off is worth it.

The next step is to make sure the `-Werror` only gets enabled in the ddtrace CI.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -37,6 +37,7 @@ test_containers:
     GRPC_RUBY_BUILD_PROCS: 6
     DD_INSTRUMENTATION_TELEMETRY_ENABLED: false
     TEST_REDIS_OLD_HOST: redis_old
+    DD_PROFILING_CI: true
   - &container_parameters_environment
     - *container_base_environment
     - TEST_DATADOG_INTEGRATION: 1

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -37,7 +37,7 @@ test_containers:
     GRPC_RUBY_BUILD_PROCS: 6
     DD_INSTRUMENTATION_TELEMETRY_ENABLED: false
     TEST_REDIS_OLD_HOST: redis_old
-    DD_PROFILING_CI: true
+    DDTRACE_CI: true
   - &container_parameters_environment
     - *container_base_environment
     - TEST_DATADOG_INTEGRATION: 1

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -16,7 +16,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
+    environment: &common-environment
       - BUNDLE_GEMFILE=/app/Gemfile
       - DD_AGENT_HOST=ddagent
       - TEST_DATADOG_INTEGRATION=1
@@ -52,20 +52,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -88,20 +75,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -124,20 +98,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -160,20 +121,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -196,20 +144,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -232,20 +167,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -267,7 +189,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
+    environment: &common-environment-3x
       - BUNDLE_GEMFILE=/app/Gemfile
       - DD_AGENT_HOST=ddagent
       - TEST_DATADOG_INTEGRATION=1
@@ -299,18 +221,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment-3x
     stdin_open: true
     tty: true
     volumes:
@@ -332,18 +243,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment-3x
     stdin_open: true
     tty: true
     volumes:
@@ -364,18 +264,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment-3x
     stdin_open: true
     tty: true
     volumes:
@@ -397,18 +286,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment-3x
     stdin_open: true
     tty: true
     volumes:
@@ -429,18 +307,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment-3x
     stdin_open: true
     tty: true
     volumes:
@@ -465,20 +332,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -500,20 +354,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -535,20 +376,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:
@@ -571,20 +399,7 @@ services:
       - redis
       - redis_old
     env_file: ./.env
-    environment:
-      - BUNDLE_GEMFILE=/app/Gemfile
-      - DD_AGENT_HOST=ddagent
-      - TEST_DATADOG_INTEGRATION=1
-      - TEST_ELASTICSEARCH_HOST=elasticsearch
-      - TEST_MEMCACHED_HOST=memcached
-      - TEST_MONGODB_HOST=mongodb
-      - TEST_MYSQL_HOST=mysql
-      - TEST_POSTGRES_HOST=postgres
-      - TEST_PRESTO_HOST=presto
-      - TEST_PRESTO_PORT=8080
-      - TEST_REDIS_HOST=redis
-      - TEST_REDIS_OLD_HOST=redis_old
-      - TEST_REDIS_OLD_PORT=6379
+    environment: *common-environment
     stdin_open: true
     tty: true
     volumes:

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -30,6 +30,7 @@ services:
       - TEST_REDIS_HOST=redis
       - TEST_REDIS_OLD_HOST=redis_old
       - TEST_REDIS_OLD_PORT=6379
+      - DD_PROFILING_CI=true
     stdin_open: true
     tty: true
     volumes:

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -30,7 +30,7 @@ services:
       - TEST_REDIS_HOST=redis
       - TEST_REDIS_OLD_HOST=redis_old
       - TEST_REDIS_OLD_PORT=6379
-      - DD_PROFILING_CI=true
+      - DDTRACE_CI=true
     stdin_open: true
     tty: true
     volumes:

--- a/ext/ddtrace_profiling_loader/extconf.rb
+++ b/ext/ddtrace_profiling_loader/extconf.rb
@@ -26,7 +26,7 @@ end
 
 # Because we can't control what compiler versions our customers use, shipping with -Werror by default is a no-go.
 # But we can enable it in CI, so that we quickly spot any new warnings that just got introduced.
-add_compiler_flag '-Werror' if ENV['CI'] == 'true'
+add_compiler_flag '-Werror' if ENV['DD_PROFILING_CI'] == 'true'
 
 # Older gcc releases may not default to C99 and we need to ask for this. This is also used:
 # * by upstream Ruby -- search for gnu99 in the codebase

--- a/ext/ddtrace_profiling_loader/extconf.rb
+++ b/ext/ddtrace_profiling_loader/extconf.rb
@@ -26,7 +26,7 @@ end
 
 # Because we can't control what compiler versions our customers use, shipping with -Werror by default is a no-go.
 # But we can enable it in CI, so that we quickly spot any new warnings that just got introduced.
-add_compiler_flag '-Werror' if ENV['DD_PROFILING_CI'] == 'true'
+add_compiler_flag '-Werror' if ENV['DDTRACE_CI'] == 'true'
 
 # Older gcc releases may not default to C99 and we need to ask for this. This is also used:
 # * by upstream Ruby -- search for gnu99 in the codebase

--- a/ext/ddtrace_profiling_native_extension/extconf.rb
+++ b/ext/ddtrace_profiling_native_extension/extconf.rb
@@ -79,7 +79,7 @@ end
 
 # Because we can't control what compiler versions our customers use, shipping with -Werror by default is a no-go.
 # But we can enable it in CI, so that we quickly spot any new warnings that just got introduced.
-add_compiler_flag '-Werror' if ENV['CI'] == 'true'
+add_compiler_flag '-Werror' if ENV['DD_PROFILING_CI'] == 'true'
 
 # Older gcc releases may not default to C99 and we need to ask for this. This is also used:
 # * by upstream Ruby -- search for gnu99 in the codebase

--- a/ext/ddtrace_profiling_native_extension/extconf.rb
+++ b/ext/ddtrace_profiling_native_extension/extconf.rb
@@ -79,7 +79,7 @@ end
 
 # Because we can't control what compiler versions our customers use, shipping with -Werror by default is a no-go.
 # But we can enable it in CI, so that we quickly spot any new warnings that just got introduced.
-add_compiler_flag '-Werror' if ENV['DD_PROFILING_CI'] == 'true'
+add_compiler_flag '-Werror' if ENV['DDTRACE_CI'] == 'true'
 
 # Older gcc releases may not default to C99 and we need to ask for this. This is also used:
 # * by upstream Ruby -- search for gnu99 in the codebase

